retroactive/src/utils.ts

```diff
@@ -0,0 +1,160 @@
+import { BigNumber } from "ethers";
+import { last } from "lodash";
+
+const ZERO_ADDRESS = "0x0000000000000000000000000000000000000000";
+
+type BlockHistory = {
+  blockNumber: number;
+  amount: BigNumber;
+};
+
+type AccountHistory = {
+  [address: string]: BlockHistory[];
+};
+
+type AccountReward = {
+  [address: string]: BigNumber;
+};
+
+const bigNumberify = (value): BigNumber => {
+  if (BigNumber.isBigNumber(value)) {
+    return value;
+  } else {
+    return BigNumber.from(value);
+  }
+};
+
+// Calculate a reward score based on the holdings of a token multipled by the
+// number of blocks it was held for.
+const cumulativeRewardScore = (
```

Comment:
I'm reasonably sure there's a bug in this method. Looking at the score data, swap related contract are relatively "outperforming" holders.


I don't know if this is the same bug or not, but contract 0xa79828df1850e8a3a3064576f380d90aecdd3359 is a deposit contract that never actually holds OUSD outside of an in-out transaction. And yet it's marked as the 7th biggest receiver of an airdrop.

<img width="1267" alt="image" src="https://user-images.githubusercontent.com/837/174835109-ec00b400-3b46-4afe-ad69-b23e0b4df60e.png">

Reply:
Somehow it's calculating every single balance forward as if it's held all the way to the snapshot block.

<img width="636" alt="image" src="https://user-images.githubusercontent.com/837/174840564-8cfdfc06-0330-432c-a699-f90aa5ef7935.png">

<img width="349" alt="image" src="https://user-images.githubusercontent.com/837/174840702-8fdc800a-5c7f-476c-b5eb-c61b375f611e.png">

Reply:
Bingo: 
<img width="591" alt="image" src="https://user-images.githubusercontent.com/837/174841153-a229caac-077d-4eaf-99f7-ef894bed4fd4.png">

Reply:
Also need to drop skipping the first entry and change the non-end block to something more like

```javascript
lastBlock = blockNumber;
firstBlock = Math.min(stopBlock, history[currentIndex + 1].blockNumber)
```

Reply:
I think it works if we remove skipping the first history entry and leave the rest as is.

`Math.min(stopBlock, history[currentIndex + 1].blockNumber)` should always just resolve to stopBlock because the history stops at stopBlock (i.e. we don't query events passed that point.

Am I thinking about that correctly? 😅

Reply:
Doesn't feel like that's correct.

Reply:
Looks roughly good now.